File: Client/Gamify.Client.Net/Gamify.Client.Net/Managers/MessageManager.cs
using Gamify.Client.Net.Client;
using Gamify.Client.Net.Services;
using Gamify.Contracts.Notifications;
using Gamify.Contracts.Requests;
using System;

namespace Gamify.Client.Net.Managers
{
    public class MessageManager
    {
        private readonly IGameService<MessageRequestObject, MessageNotificationObject> messageService;
        private readonly IGameService<TypingMessageRequestObject, TypingMessageNotificationObject> typingMessageService;
        private readonly IGameListener<ErrorNotificationObject> errorListener;

        public event EventHandler<GameNotificationEventArgs<MessageNotificationObject>> MessageNotificationReceived;
        public event EventHandler<GameNotificationEventArgs<TypingMessageNotificationObject>> TypingMessageNotificationReceived;
        public event EventHandler<GameNotificationEventArgs<ErrorNotificationObject>> ErrorNotificationReceived;

        public MessageManager(string playerName, IGameClientFactory clientFactory)
        {
            var gameClient = clientFactory.GetGameClient(playerName);

            this.messageService = new GameService<MessageRequestObject, MessageNotificationObject>(GameRequestType.Message, GameNotificationType.Message, gameClient);
            this.typingMessageService = new GameService<TypingMessageRequestObject, TypingMessageNotificationObject>(GameRequestType.TypingMessage, GameNotificationType.TypingMessage, gameClient);
            this.errorListener = new GameListener<ErrorNotificationObject>(GameNotificationType.Error, gameClient);

            this.messageService.NotificationReceived += (sender, args) =>
            {
                this.NotifyMessage(args);
            };
            this.typingMessageService.NotificationReceived += (sender, args) =>
            {
                this.NotifyTypingMessage(args);
            };
            this.errorListener.NotificationReceived += (sender, args) =>
            {
                this.NotifyError(args);
            };
        }

        public void SendMessage(MessageRequestObject messageRequest)
        {
            this.messageService.Send(messageRequest);
        }

        public void SendTypingMessage(TypingMessageRequestObject typingMessageRequest)
        {
            this.typingMessageService.Send(typingMessageRequest);
        }

        private void NotifyMessage(GameNotificationEventArgs<MessageNotificationObject> args)
        {
            if (this.MessageNotificationReceived != null)
            {
                this.MessageNotificationReceived(this, args);
            }
        }

        private void NotifyTypingMessage(GameNotificationEventArgs<TypingMessageNotificationObject> args)
        {
            if (this.TypingMessageNotificationReceived != null)
            {
                this.TypingMessageNotificationReceived(this, args);
            }
        }

        private void NotifyError(GameNotificationEventArgs<ErrorNotificationObject> args)
        {
            if (this.ErrorNotificationReceived != null)
            {
                this.ErrorNotificationReceived(this, args);
            }
        }
    }
}

File: Java/gamify-java-client/src/main/java/ar/com/mdf/gamify/managers/MessageManager.java
package ar.com.mdf.gamify.managers;

import java.net.URISyntaxException;

import ar.com.mdf.gamify.client.IGamifyClient;
import ar.com.mdf.gamify.enums.NotificationType;
import ar.com.mdf.gamify.enums.RequestType;
import ar.com.mdf.gamify.exceptions.NotConnectedException;
import ar.com.mdf.gamify.factory.IClientFactory;
import ar.com.mdf.gamify.objects.notification.ErrorNotificationObject;
import ar.com.mdf.gamify.objects.notification.MessageNotificationObject;
import ar.com.mdf.gamify.objects.request.MessageRequestObject;
import ar.com.mdf.gamify.observers.IGameObserver;
import ar.com.mdf.gamify.services.GameListener;
import ar.com.mdf.gamify.services.GameService;
import ar.com.mdf.gamify.services.IGameListener;
import ar.com.mdf.gamify.services.IGameService;

public class MessageManager {
	private IGameService<MessageRequestObject, MessageNotificationObject> messageService;
	private IGameService<MessageRequestObject, MessageNotificationObject> typingMessageService;
	private IGameListener<ErrorNotificationObject> errorListener;
	private IGamifyClient client;
	
	public MessageManager(IClientFactory clientFactory) throws NotConnectedException {
		try {
			this.client = clientFactory.getClient();
		} catch (URISyntaxException e) {
			throw new NotConnectedException(e);
		}
		this.messageService = new GameService<MessageRequestObject, MessageNotificationObject>(RequestType.Message, NotificationType.Message, this.client, MessageNotificationObject.class);
		this.typingMessageService = new GameService<MessageRequestObject, MessageNotificationObject>(RequestType.TypingMessage, NotificationType.TypingMessage, this.client, MessageNotificationObject.class);
		this.errorListener = new GameListener<ErrorNotificationObject>(NotificationType.Error, ErrorNotificationObject.class);
	}
	
	public void suscribeMessage(IGameObserver<MessageNotificationObject> observer){
		this.messageService.suscribe(observer);
	}
	public void suscribeTypingMessage(IGameObserver<MessageNotificationObject> observer){
		this.typingMessageService.suscribe(observer);
	}
	public void suscribeError(IGameObserver<ErrorNotificationObject> observer){
		this.errorListener.suscribe(observer);
	}
	
	public void sendMessage(MessageRequestObject openGameRequest) {
		this.messageService.send(openGameRequest);
	}
	public void sendTypingMessage(MessageRequestObject openGameRequest) {
		this.typingMessageService.send(openGameRequest);
	}
}
